reference_guide/custom_language_support/declarations_and_references_20203.md

```diff
@@ -0,0 +1,92 @@
+---
+title: Declarations and References
+---
+<!-- Copyright 2000-2020 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file. -->
+
+> **WARNING** This API is currently in development and thus in experimental state
+
+## Declarations
+
+Each symbol may be declared in zero or more places, for example:
+- JVM package is a symbol with several declarations (split packages);
+- C# partial class is a symbol with several declarations;
+- property key is a symbol declared in several files simultaneously;
+- Java local variable is a symbol with a single declaration;
+- file is a symbol without declarations, it has only references.
+
+The declarations are implementations of
+[`SymbolDeclaration`](upsource:///platform/core-api/src/com/intellij/model/SymbolDeclaration.java). 
+The declarations in PSI elements are implementations of
+[`PsiSymbolDeclaration`](upsource:///platform/core-api/src/com/intellij/model/psi/PsiSymbolDeclaration.java).
+
+To report a declaration in a PSI element either:
+- implement and register 
+  [`PsiSymbolDeclarationProvider`](upsource:///platform/core-api/src/com/intellij/model/psi/PsiSymbolDeclarationProvider.java);
+- implement `PsiSymbolDeclaration` directly in the `PsiElement`.
+
+
+## References
+
+The references are implementations of 
+[`SymbolReference`](upsource:///platform/core-api/src/com/intellij/model/SymbolReference.java) interface. 
+The references from PSI elements are implementations of 
+[`PsiSymbolReference`](upsource:///platform/core-api/src/com/intellij/model/psi/PsiSymbolReference.java) interface.
+
+The main method of the `SymbolReference` interface is `resolveReference()`, which returns the collection of results 
+(symbols to which the reference points plus additional data), or an empty collection if it was not possible to resolve the reference 
+(for example, should it point to an undefined class). A counterpart to the `resolveReference()` method is `resolvesTo()`, 
+which checks if the reference resolves to the specified element. The `resolvesTo()` method can be implemented 
+to perform the tree walk only if the element text is equal to the text of the reference.
+
+For convenience, if the reference can possibly be resolved:
+- with a single result, then it might be extended from 
+[`SingleResultReference`](upsource:///platform/core-api/src/com/intellij/model/SingleResultReference.java);
+- to a single symbol without additional data, then it might be extended from 
+[`SingleTargetReference`](upsource:///platform/core-api/src/com/intellij/model/SingleTargetReference.java);
+- to multiple symbols without additional data, then 
+[`SymbolResolveResult#fromSymbol()`](upsource:///platform/core-api/src/com/intellij/model/SymbolResolveResult.java) might be used.
+
+
+### Own references
+
+The own references are the references found in PSI elements which are considered as references by the language.
```

Comment:
"The own references" seems ungrammatical

Reply:
Maybe singular instead of plural (and keep "the")?

Reply:
I'd be indefinite then. An own reference is a reference found ....